File: go/1855.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	var c, r int
	fmt.Scan(&c, &r)

	var grid [][]rune
	for i := 0; i < r; i++ {
		line, _ := reader.ReadString('\n')
		grid = append(grid, []rune(strings.TrimSpace(line)))
	}

	visited := make(map[string]bool)
	direction := ""
	current := [2]int{0, 0}

	getCurrent := func(curr [2]int) rune {
		return grid[curr[0]][curr[1]]
	}

	for !visited[fmt.Sprintf("%v", current)] && getCurrent(current) != '*' {
		visited[fmt.Sprintf("%v", current)] = true
		currentCell := getCurrent(current)
		if currentCell == '>' || currentCell == '<' || currentCell == 'v' || currentCell == '^' {
			direction = string(currentCell)
		}

		switch direction {
		case ">":
			if current[1]+1 >= c {
				break
			}
			current[1]++
		case "<":
			if current[1]-1 < 0 {
				break
			}
			current[1]--
		case "v":
			if current[0]+1 >= r {
				break
			}
			current[0]++
		case "^":
			if current[0]-1 < 0 {
				break
			}
			current[0]--
		}
	}

	if getCurrent(current) != '*' {
		fmt.Println("!")
	} else {
		fmt.Println("*")
	}
}

File: python/1855.py
from collections import defaultdict

c = int(input())
r = int(input())

map = []

for idx in range(r):
    map.append(list(input()))

visited = defaultdict(int)

direction = ""
current: tuple[int, int] = (0, 0)


def current_p(map, curr):
    return map[curr[0]][curr[1]]


while not visited[str(current)] and current_p(map, current) != "*":
    visited[str(current)] = 1
    if current_p(map, current) in [">", "<", "v", "^"]:
        direction = current_p(map, current)

    if direction == ">":
        if current[1] + 1 >= c:
            break
        current = [current[0], current[1] + 1]
    elif direction == "<":
        if current[1] - 1 < 0:
            break
        current = [current[0], current[1] - 1]
    elif direction == "v":
        if current[0] + 1 >= r:
            break
        current = [current[0] + 1, current[1]]
    elif direction == "^":
        if current[0] - 1 < 0:
            break
        current = [current[0] - 1, current[1]]

if current_p(map, current) != "*":
    print("!")
else:
    print("*")
